File: tools/Assemblers/js/Assembler.js
const INSTRUCTIONS = {
    A: "ADDRESS",
    C: "COMPUTE",
    L: "LABEL",
};

class Assembler {
    // we start storing variables at 16th register. From M[16] to M[256].
    freeVariableAddress = 16;

    VARIABLES = {
        // pre-defined vars
        R0: 0,
        R1: 1,
        R2: 2,
        R3: 3,
        R4: 4,
        R5: 5,
        R6: 6,
        R7: 7,
        R8: 8,
        R9: 9,
        R10: 10,
        R11: 11,
        R12: 12,
        R13: 13,
        R14: 14,
        R15: 15,
        // IO
        SCREEN: 16384,
        KBD: 24578,
        // VM
        SP: 0,
        LCL: 1,
        ARG: 2,
        THIS: 3,
        THAT: 4,
    };

    constructor() {
        return this;
    }

    assemble(asmFile) {
        const noComments = (line) =>
            (line.includes("//")
                ? line.slice(0, line.indexOf("//"))
                : line
            ).trim();
        const noWhitespaces = (line) => Boolean(line);
        const noLabels = (line) => this._getType(line) !== INSTRUCTIONS.L;
        const inLines = "\r\n";

        const asmCode = asmFile
            .split(inLines)
            .map(noComments)
            .filter(noWhitespaces);

        asmCode.forEach(this._initLabels.bind(this));

        const hackCode = asmCode
            .filter(noLabels)
            .map(this._asm2bin.bind(this))
            .join(inLines);

        return hackCode;
    }

    LABELS = {};
    labelLineNumber = 0;
    _initLabels(instruction) {
        if (
            this._getType(instruction) === INSTRUCTIONS.A ||
            this._getType(instruction) === INSTRUCTIONS.C
        )
            this.labelLineNumber++;
        else if (this._getType(instruction) === INSTRUCTIONS.L) {
            // eslint-disable-next-line no-unused-vars
            const [_, value] = instruction.match(/\((.*)\)/i); // todo: don't use regexp
            if (!this.LABELS[value]) this.LABELS[value] = this.labelLineNumber;
        }
    }

    _asm2bin(instruction) {
        switch (this._getType(instruction)) {
            case INSTRUCTIONS.A:
                return this._translateA(instruction);
            case INSTRUCTIONS.C:
                return this._translateC(instruction);
        }
    }

    _getType(instruction) {
        if (instruction.includes("@")) return INSTRUCTIONS.A;
        else if (instruction.startsWith("(") && instruction.endsWith(")"))
            return INSTRUCTIONS.L;
        else if (instruction.match(/A?M?D?=/) || instruction.match(/;J/))
            return INSTRUCTIONS.C; // todo: add strong if statement predicate
    }

    _translateA(instruction) {
        const decToBin = (dec) => (+dec).toString(2);
        const fillWith15Bits = (bin) =>
            "000000000000000".slice(bin.length) + bin;
        const getValueOfInstructionA = (instruction) => instruction.slice(1);

        let value = getValueOfInstructionA(instruction);

        const isLabelOrVariable = value.match(/\D+/i);

        if (isLabelOrVariable)
            value =
                this._getLabel(value) ||
                this._getVariable(value) ||
                this._initVariable(value);

        const hack = `0${fillWith15Bits(decToBin(value))}`;

        return hack;
    }

    _getLabel(value) {
        return this.LABELS[value] !== undefined && String(this.LABELS[value]);
    }

    _getVariable(value) {
        return (
            this.VARIABLES[value] !== undefined && String(this.VARIABLES[value])
        );
    }

    _initVariable(value) {
        this.VARIABLES[value] = this.freeVariableAddress;
        this.freeVariableAddress++;
        return String(this.VARIABLES[value]);
    }

    _translateC(instruction) {
        const { comp, dest, jump } = this._separateC(instruction);

        this._validateC({ comp, dest, jump }, instruction);

        return `111${comp}${dest}${jump}`;
    }

    _separateC(instruction) {
        const [dstcmp, jmp = null] = instruction.split(";");
        const [dst, cmp] = dstcmp.includes("=")
            ? dstcmp.split("=")
            : [null, dstcmp];

        const comp = this._comp(cmp);
        const dest = this._dest(dst);
        const jump = this._jump(jmp);

        return { comp, dest, jump };
    }

    _validateC({ comp, dest, jump }, instruction) {
        if (comp === null)
            throw Error(`unknown comp in instruction: "${instruction}"`);
        if (dest === null)
            throw Error(`unknown dest in instruction "${instruction}"`);
        if (jump === null)
            throw Error(`unknown jump in instruction "${instruction}"`);
    }

    _dest(symbols) {
        // if (symbols && symbols.length >= 2) console.log({ symbols });
        // prettier-ignore
        switch (symbols) {
            case "M":   return '001';
            case "D":   return '010';
            case "MD":  return '011';
            case "A":   return '100';
            case "AM":  return '101';
            case "AD":  return '110';
            case "AMD": return '111';
            case null:  return '000';
            default:    return null;
        }
    }

    _comp(symbols) {
        // prettier-ignore
        switch (symbols) {
            // a = 0
            case "0":   return '0101010';
            case "1":   return '0111111';
            case "-1":  return '0111010';
            case "D":   return '0001100';
            case "A":   return '0110000';
            case "!D":  return '0001101';
            case "!A":  return '0110001';
            case "-D":  return '0001111';
            case "-A":  return '0110011';
            case "D+1": return '0011111';
            case "A+1": return '0110111';
            case "D-1": return '0001110';
            case "A-1": return '0110010';
            case "D+A": return '0000010';
            case "D-A": return '0010011';
            case "A-D": return '0000111';
            case "D&A": return '0000000';
            case "D|A": return '0010101';
            // a = 1
            case "M":   return '1110000';
            case "!M":  return '1110001';
            case "-M":  return '1110011';
            case "M+1": return '1110111';
            case "M-1": return '1110010';
            case "D+M": return '1000010';
            case "D-M": return '1000010';
            case "M-D": return '1000111';
            case "D&M": return '1000000';
            case "D|M": return '1010101';
            case null:  return '';
            default:    return null;
        }
    }

    _jump(symbols) {
        // prettier-ignore
        switch (symbols) {
            case "JGT":     return '001';
            case "JEQ":     return '010';
            case "JGE":     return '011';
            case "JLT":     return '100';
            case "JNE":     return '101';
            case "JLE":     return '110';
            case "JMP":     return '111';
            case null:      return '000';
            default:        return null;
        }
    }
}

module.exports = Assembler;

File: tools/Assemblers/python/assembler.py
#!/usr/bin/env python

import re
import sys
from enum import Enum, auto
from pathlib import Path
from typing import Mapping

COMPS = {
    "0": "0101010",
    "1": "0111111",
    "-1": "0111010",
    "D": "0001100",
    "A": "0110000",
    "M": "1110000",
    "!D": "0001101",
    "!A": "0110001",
    "!M": "1110001",
    "-D": "0001111",
    "-A": "0110011",
    "-M": "1110011",
    "D+1": "0011111",
    "A+1": "0110111",
    "M+1": "1110111",
    "D-1": "0001110",
    "A-1": "0110010",
    "M-1": "1110010",
    "D+A": "0000010",
    "D+M": "1000010",
    "D-A": "0010011",
    "D-M": "1010011",
    "A-D": "0000111",
    "M-D": "1000111",
    "D&A": "0000000",
    "D&M": "1000000",
    "D|A": "0010101",
    "D|M": "1010101",
}

JUMPS = {
    # None: '000',
    "JGT": "001",
    "JEQ": "010",
    "JGE": "011",
    "JLT": "100",
    "JNE": "101",
    "JLE": "110",
    "JMP": "111",
}

COMMENT_RE = re.compile(r"//.*$")
C_COMMAND_RE = re.compile(
    "^(?P<dest>A?M?D?=|)"
    + ("(?P<comp>" + "|".join(re.escape(comp) for comp in COMPS) + ")")
    + ("(?:;(?P<jump>" + "|".join(re.escape(jump) for jump in JUMPS) + "))?$")
)


class Command(Enum):
    A_COMMAND = auto()
    C_COMMAND = auto()
    L_COMMAND = auto()


def parse_commands(asm_file: Path):
    with asm_file.open() as f:
        for line_no, line in enumerate(f):
            line = COMMENT_RE.sub("", line).strip()
            if line:
                if line.startswith("(") and line.endswith(")"):
                    yield (Command.L_COMMAND, line[1:-1])

                elif line.startswith("@"):
                    yield (Command.A_COMMAND, line[1:])

                elif C_COMMAND_RE.fullmatch(line):
                    yield (Command.C_COMMAND, line)

                else:
                    raise RuntimeError(f"Unable to parse {asm_file}:{line_no}: {line}")


def generate_symbols(asm_file: Path):
    symbol_table = {
        **{name: n for n, name in enumerate(["SP", "LCL", "ARG", "THIS", "THAT"])},
        **{f"R{n}": n for n in range(16)},
        "SCREEN": 0x4000,
        "KBD": 0x6000,
    }

    cur_address = 0
    for command, line in parse_commands(asm_file):
        if (command == Command.C_COMMAND) or (command == Command.A_COMMAND):
            cur_address += 1
        elif command == Command.L_COMMAND:
            if line not in symbol_table:
                symbol_table[line] = cur_address

    return symbol_table


class DynamicSymbolTable(dict):
    next_variable_address = 16

    def __missing__(self, symbol):
        self[symbol] = self.next_variable_address
        self.next_variable_address += 1
        return self[symbol]


def assemble(asm_file: Path, symbols: Mapping):
    symbols = DynamicSymbolTable(symbols)
    with asm_file.with_suffix(".hack").open("w") as hack_file:
        for command, line in parse_commands(asm_file):
            if command == Command.A_COMMAND:
                try:
                    value = int(line)
                except ValueError:
                    value = symbols[line]
                hack_file.write(f"0{value:015b}\n")
            elif command == Command.C_COMMAND:
                match = C_COMMAND_RE.fullmatch(line)
                if match:
                    value = COMPS[match["comp"]]
                    for dest in "ADM":
                        if dest in match["dest"]:
                            value += "1"
                        else:
                            value += "0"
                    value += JUMPS.get(match["jump"], "000")
                    hack_file.write(f"111{value}\n")


def main(filename: str):
    asm_file = Path(filename)
    symbols = generate_symbols(asm_file)
    assemble(asm_file, symbols)


if __name__ == "__main__":
    main(sys.argv[1])
